### DISCUSSION TITLE
Unable to get Replex working on local devices -- works remotely!

### DISCUSSION THREAD

**Discussion by adamblaine** [lostb1t/replex#106] [Q&A]
Hello, I'm hoping someone here can help me get this figured out. I followed the instructions on the readme and poured over the reddit threads I could find looking for a solution, but haven't found one yet. 

I am able to get Replex working and looking correct on devices outside my local network, but when I try to connect from my server pc, the smart tv apps via wifi, or my phone app, it looks like the default layout (no merged collections, no hero banners). I can disconnect my phone from wifi and open plex and the replex proxy works! But when I'm connected -- no such luck.

For reference: I'm running Plex on a windows machine as a service. I'm running Replex via Docker, which is also running various other services including a reverse proxy (swag). This is my docker-compose for reference:
```dockerfile
   replex:
    image: ghcr.io/lostb1t/replex:latest
    container_name: replex
    environment:
      REPLEX_HOST: http://host.docker.internal:32400
      # REPLEX_HERO_ROWS: movie.recentlyadded,tv.recentlyadded
      REPLEX_INCLUDE_WATCHED: false
      REPLEX_FORCE_MAXIMUM_QUALITY: true
      REPLEX_DISABLE_CONTINUE_WATCHING: false
      # REPLEX_CACHE_TTL: 1800
      RUST_LOG: info,replex=debug
      # REPLEX_CACHE_ROWS: false
      # REPLEX_CACHE_ROWS_REFRESH: true
    ports:
      - 3001:80
    restart: always
```
My reverse proxy config includes the following:
```dockerfile
server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    server_name plex.*;
 
    include /config/nginx/ssl.conf;
    client_max_body_size 0;
    #proxy_redirect off;
    #proxy_buffering off;
    
    location / {
        include /config/nginx/proxy.conf;
        include /config/nginx/resolver.conf;
        proxy_pass http://host.docker.internal:3001;
    }
    location /video/:/transcode/universal/session {
        proxy_pass http://host.docker.internal:32400;
    }
    location /library/parts {
        proxy_pass http://host.docker.internal:32400;  
    }
}
```
When checking the plex connections in the xml, I'm seeing these:
```dockerfile
<connections>
<connection protocol="https" address="plex.mydomainname.net" port="443" uri="https://plex.mydomainname.net:443" local="0" relay="0" IPv6="0"/>
<connection protocol="http" address="192.168.0.121" port="32400" uri="http://192.168.0.121:32400" local="1" relay="0" IPv6="0"/>
</connections>
```
The connection through my domain via reverse proxy seems to be connecting correctly. But there's another connection through this machine's internal IP address that seems to be bypassing the proxy. There are also other various devices on the network (the TVs, phones, etc) which are connecting locally. 

Within Plex, I've disabled the remote access and under Network I've disabled GDM and set the Custom Server Access URLs as:
https://plex.mydomainname.net:443

Within my Replex container, the debug logs look mostly like this:
```dockerfile
2023-11-21 12:19:32 2023-11-21T20:19:32.005366Z  INFO replex: Replex version 0.19.5
2023-11-21 12:19:32 2023-11-21T20:19:32.043190Z  INFO salvo_core::server: listening [HTTP/1.1] on http://0.0.0.0:80
2023-11-21 12:22:15 2023-11-21T20:22:15.643011Z DEBUG replex::logging: Response status=200 OK duration=57.923025ms
2023-11-21 12:22:16 2023-11-21T20:22:16.116253Z DEBUG replex::logging: Response status=200 OK duration=11.400205ms
```

But when I try to navigate to https://plex.mydomainname.net from a local machine, I see this in the log:
```dockerfile
2023-11-21 13:43:15 2023-11-21T21:43:15.609882Z DEBUG replex::logging: Response status=401 Unauthorized duration=14.428108ms
2023-11-21 13:43:15 2023-11-21T21:43:15.613933Z DEBUG replex::logging: Response status=200 OK duration=30.089917ms
2023-11-21 13:43:15 2023-11-21T21:43:15.616491Z DEBUG replex::logging: Response status=401 Unauthorized duration=12.870708ms
2023-11-21 13:43:15 2023-11-21T21:43:15.656261Z DEBUG replex::logging: Response status=200 OK duration=326.666088ms
2023-11-21 13:43:15 2023-11-21T21:43:15.740786Z DEBUG replex::logging: Response status=200 OK duration=14.526208ms
2023-11-21 13:43:15 2023-11-21T21:43:15.748496Z DEBUG replex::logging: Response status=200 OK duration=15.133209ms
2023-11-21 13:43:15 2023-11-21T21:43:15.755911Z DEBUG replex::logging: Response status=200 OK duration=15.752409ms
2023-11-21 13:43:15 2023-11-21T21:43:15.766531Z DEBUG replex::logging: Response status=101 Switching Protocols duration=17.56011ms
```
And it looks like the default plex layout. Any ideas as to what I've set incorrectly? Would any other information be helpful in solving this issue?

I do also have a VPN on this machine, but it also runs as a windows application outside of Docker, and I have split-tunneling enabled so that VPN bypasses Plex.

**Comment by lostb1t** [comment]
https://github.com/lostb1t/replex/discussions/83#discussioncomment-6992653

**Reply by adamblaine** [reply]
Thanks for the reply. I had already disabled GDM and I'm running Plex in windows, not in docker. 

I'm not sure how to go about isolating Plex from my local network outside of throwing it behind my VPN?

**Reply by lostb1t** [reply]
> If that doesn't work you can use this url
> 
> https://clients.plex.tv/api/v2/resources?includeIPv6=1&includeRelay=1&X-Plex-Language=en-NL&X-Plex-Token=YOURTOKEN&X-Plex-Client-Identifier=1234
> 
> (Chance token to your token)
> 
> To see I plex reports the correct server address to your clients

**Reply by lostb1t** [reply]
try blocking port 32400 with a firewall on your plex machine for external access (replex should ofcourse still reach it but if its the same machine it should work

**Comment by adamblaine** [comment]
> > If that doesn't work you can use this url
> > https://clients.plex.tv/api/v2/resources?includeIPv6=1&includeRelay=1&X-Plex-Language=en-NL&X-Plex-Token=YOURTOKEN&X-Plex-Client-Identifier=1234
> > (Chance token to your token)
> > To see I plex reports the correct server address to your clients

This is the output on that page (with edits to the domain name and external IP):
```dockerfile
<resources>
<resource name="Host-PC" product="Plex Media Server" productVersion="1.32.8.7639-fb6452ebf" platform="Windows" platformVersion="10.0 (Build 22631)" device="PC" clientIdentifier="ad634952a7a892da0efc3216e137607a8cace59e" createdAt="2017-12-07T09:24:46Z" lastSeenAt="2023-11-22T16:52:55Z" provides="server" ownerId="" sourceTitle="" publicAddress="MYEXTERNALIP" accessToken="REMOVEDTOKEN" owned="1" home="0" synced="0" relay="0" presence="1" httpsRequired="0" publicAddressMatches="1" dnsRebindingProtection="0" natLoopbackSupported="0">
<connections>
<connection protocol="https" address="plex.MYDOMAINNAME.net" port="443" uri="https://plex.MYDOMAINNAME.net:443" local="0" relay="0" IPv6="0"/>
<connection protocol="http" address="192.168.0.121" port="32400" uri="http://192.168.0.121:32400" local="1" relay="0" IPv6="0"/>
</connections>
</resource>
<resource name="LG OLED65C8PUA" product="Plex for LG" productVersion="5.72.3" platform="webOS" platformVersion="4.4.0" device="webOS 4.4.0" clientIdentifier="iotipbv9jbbmh6tbyw8ho8xp" createdAt="2022-08-27T16:16:17Z" lastSeenAt="2023-11-22T07:57:37Z" provides="client,player" ownerId="" sourceTitle="" publicAddress="MYEXTERNALIP" accessToken="REMOVEDTOKEN" owned="1" home="0" synced="0" relay="0" presence="0" httpsRequired="0" publicAddressMatches="1">
<connections>
<connection protocol="https" address="plex.MYDOMAINNAME.net" port="443" uri="https://plex.MYDOMAINNAME.net:443" local="0" relay="0" IPv6="0"/>
<connection protocol="http" address="192.168.0.121" port="32400" uri="http://192.168.0.121:32400" local="1" relay="0" IPv6="0"/>
</connections>
</resource>
<resource name="Galaxy S22 Ultra" product="Plex for Android (Mobile)" productVersion="10.3.0.4685" platform="Android" platformVersion="13" device="SM-S908U1" clientIdentifier="b93a5e08a85e505e-com-plexapp-android" createdAt="2022-09-03T03:34:23Z" lastSeenAt="2023-11-21T17:41:35Z" provides="player,pubsub-player,controller,sync-target" ownerId="" sourceTitle="" publicAddress="MYEXTERNALIP" accessToken="" owned="1" home="0" synced="0" relay="0" presence="0" httpsRequired="0" publicAddressMatches="1">
<connections>
<connection protocol="http" address="192.168.0.11" port="32500" uri="http://192.168.0.11:32500" local="1" relay="0" IPv6="0"/>
</connections>
</resource>
<resource name="Android" product="Plexamp" productVersion="4.8.4" platform="Android" platformVersion="13" device="Android" clientIdentifier="78e959b1-edd6-47e4-af2f-a6646228abf9" createdAt="2022-11-02T19:09:55Z" lastSeenAt="2023-11-21T22:00:06Z" provides="client,player,pubsub-player" ownerId="" sourceTitle="" publicAddress="MYEXTERNALIP" accessToken="" owned="1" home="0" synced="0" relay="0" presence="0" httpsRequired="0" publicAddressMatches="1">
<connections>
<connection protocol="http" address="192.0.0.2" port="58607" uri="http://192.0.0.2:58607" local="0" relay="0" IPv6="0"/>
<connection protocol="http" address="192.168.0.12" port="58607" uri="http://192.168.0.12:58607" local="1" relay="0" IPv6="0"/>
</connections>
</resource>
<resource name="Android" product="Plexamp" productVersion="4.8.4" platform="Android" platformVersion="13" device="Android" clientIdentifier="5ca093cc-bd67-40b6-a4a3-4958cf320662" createdAt="2023-09-02T02:00:33Z" lastSeenAt="2023-11-20T20:46:55Z" provides="client,player,pubsub-player" ownerId="" sourceTitle="" publicAddress="172.56.176.180" accessToken="" owned="1" home="0" synced="0" relay="0" presence="0" httpsRequired="0" publicAddressMatches="0">
<connections>
<connection protocol="http" address="192.168.0.6" port="50279" uri="http://192.168.0.6:50279" local="1" relay="0" IPv6="0"/>
<connection protocol="http" address="192.0.0.2" port="50279" uri="http://192.0.0.2:50279" local="0" relay="0" IPv6="0"/>
</connections>
</resource>
</resources>
``` 
Thanks again!

**Reply by lostb1t** [reply]
> try blocking port 32400 with a firewall on your plex machine for external access (replex should ofcourse still reach it but if its the same machine it should work

**Reply by lostb1t** [reply]
another thing to try is to set Secure connections in plex to required. maybe that forces it aswell. not sure tho

**Reply by adamblaine** [reply]
Thanks again for the replies. I tried blocking the port but it didn't seem to have any effect. I also set to Secure connections only but it still seemed to bypass that for local. 

I may try to have docker set up another network that I can throw it behind (even though I'm not sure how to do that as Plex isn't run from a container), but until I figure that out I put plex behind my vpn rather than allowing it to bypass it (not ideal speed-wise), and set the vpn network as the Preferred network connection in the Plex -> Network settings. 

That seems to have worked for now, but I'm going to try to figure out the virtualized network option as I don't love the idea of the VPN adding extra buffering/delays to streams.

**Comment by Zorro92** [comment]
Why not run Plex in a docker container? Then you could set it to a different port and easily block off everything else. You could probably set replex to run on port 32400 and trick your local Plex clients into thinking it's actually Plex. Seems the easiest way to go to me.

**Reply by adamblaine** [reply]
That was my first instinct, but I'm a little hesitant to convert my existing plex server from windows to docker. I'll look into it, but I've seen issues in the past where something gets set up incorrectly and the databases get corrupted and all metadata is lost.

**Reply by lostb1t** [reply]
just make a backup before doing it: https://support.plex.tv/articles/201539237-backing-up-plex-media-server-data/

**Reply by adamblaine** [reply]
Thanks -- will do. I think the other hang-up I had about doing this in the past was a difficulty in enabling hardware transcoding via docker vs using the native windows application. I'll do some googling to see if its easy to enable now.

**Reply by adamblaine** [reply]
Gave it a try but the lack of hardware transcoding made me have to go back to Plex on windows. I attempted to get it set up a couple times without success--there is just no documentation or concrete info on installing Nvidia drivers for hw transcoding on docker desktop for windows via wsl. Something about exposing the drivers (for Nvidia or Intel) is very difficult and not documented for windows docker desktop setups.

Maybe a VLAN would be an option instead of going the Plex docker route?

**Reply by adamblaine** [reply]
Just an update -- I still haven't committed to moving my plex server from windows into docker as there were a few minor annoyances with it, and I've settled with having my external plex users enjoy the replex layout while I use the standard layout on my local devices. But I did figure out the solution to getting Nvidia hardware transcoding working via docker desktop for windows. Example of my docker compose 3 here:
```dockerfile
services:
  plex:
    image: lscr.io/linuxserver/plex:latest
    container_name: plex
    network_mode: bridge
    deploy:    # <------------- Add this section
      resources:
        reservations:
          devices:
            - driver: nvidia
              #device_ids: ['0'] # this is only needed when using multiple GPUs
              count: all # number of GPUs
              capabilities: [gpu]
    # devices: 
      # - /dev/dri:/dev/dri
    ports:
      - "32400:32400"
    environment:
      - PUID
      - PGID
      - TZ=America/Los_Angeles
      - VERSION=docker
      # - PLEX_CLAIM=xxxxxxxxxxx #optional
      - NVIDIA_VISIBLE_DEVICES=all
      - NVIDIA_DRIVER_CAPABILITIES=compute,video,utility
    volumes:
      - /e/Docker/data/config/plex:/config
      - /g/TV Shows:/tv
      - /f/Movies:/movies
      - /f/Movies_4k:/movies4k
      - /f/Music/Music:/music
      - /f/Audiobooks:/audiobooks
    restart: always
``` 

I did notice a few other issues -- a family member on an external device (PS5) was having difficulty playing anything at all with the "REPLEX_FORCE_MAXIMUM_QUALITY" option set to "true."

Also, all streams in Plex Dashboard or Tautulli show as WAN devices, which is I'm assuming working properly as they're going through the proxy.

**Reply by Zorro92** [reply]
The force max quality function doesn't work well (to no fault of the developer of replex, this is because of the vastly different attributes certain Plex clients uses to communicate with the server). And yes traffic will show as wan.

**Comment by PilaScat** [comment]
Same for me, but I'm using Docker
